[Timelines] Export as CSV

VWoeltjen opened the pull request:
Adds CSV Export for timelines. Opening PR early for progress tracking.

Still to-do:

- [x] Add a Type column
- [x] Include Activity Modes (and their metadata)
- [x] Display start/end times
- [x] Progress reporting
- [x] Tests
- [x] JSDoc

VWoeltjen commented:
### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

akhenry commented on platform/features/timeline/src/actions/ExportTimelineAsCSVAction.js:
```diff
@@ -0,0 +1,70 @@
+                notification.dismiss();
+            })
+            .catch(function () {
+                notification.dismiss();
```
Notification dismissal could be done in a `finally`, but would only save one LOC here, so it doesn't really matter :)

VWoeltjen replied:
`finally` isn't an [ES6 Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) method, so was avoiding for compatibility.

That said, I think this could be equivalently rewritten as:

```
return this.task.run()
    .catch(notificationService.error.bind(notificationService, "Error exporting CSV"))
    .then(notification.dismiss.bind(notification));
```

Is this more readable or am I just playing code golf at this point?

akhenry replied:
Ah, fair enough. The existing code is absolutely fine, don't worry about it :)

akhenry commented on platform/features/timeline/src/actions/TimelineTraverser.js:
```diff
@@ -0,0 +1,85 @@
+                if (domainObject.hasCapability('composition')) {
+                    subtasks.push(addCompositionObjects());
+                }
+                if (domainObject.hasCapability('relationship')) {
```
What object relationships do timelines model? I'm just trying to understand how relationships fit into Timelines.

akhenry commented on platform/features/timeline/src/actions/TimelineCSVExporter.js:
```diff
@@ -0,0 +1,156 @@
+            var model = domainObject.getModel(),
+                compositionLength = model.composition ?
+                    model.composition.length : 0,
+                relationshipLength = (model.relationships || {}).modes ?
```
Mixing binary and ternary operators here makes it a little difficult to follow. Could either do it in two steps (define relationships, then relationshipLength, or alternatively do something like -
``` javascript
var model = domainObject.getModel(),
    composition = model.composition || [],
    modes = (model.relationships || {}).modes || [];
    ...
    maxComposition = Math.max(maxComposition, composition.length);
    maxRelationships = Math.max(maxRelationships, relationships.length);
```

I'm nitpicking though.

akhenry commented on platform/features/timeline/src/actions/ExportTimelineAsCSVTask.js:
```diff
@@ -0,0 +1,70 @@
+     * @returns {Promise} a promise that will be resolved when the
+     *          export has finished (or rejected if there are problems.)
+     */
+    ExportTimelineAsCSVTask.prototype.run = function () {
```
What's the purpose of this class? As a developer I would expect a `TimelineCSVExporter` to do the export to CSV, but instead it's done by `ExportTimelineAsCSVTask` and the `TimelineCSVExporter` actually just prepares the CSV for export. Could these be combined to reduce interface depth here?

VWoeltjen replied:
I'd separated out `TimelineTraverser` and `TimelineCSVExporter` as steps of the task that could be separated out, mainly for purposes of testability (and to keep individual units simple.) Since these aren't intended as public API I'm less concerned about number of interfaces, and would prefer to keep these separate. 

I think the naming of the steps makes this seem as odd as it does. `TimelineCSVExporter` is a pretty unclear name implying broader responsibilities - `TimelineColumnizer` maybe?

akhenry replied:
Yeah I think renaming it would help to clarify things. `TimelineColumnizer` gives a fairly clear description of what it does.

akhenry commented:
Looks great. Two comments inline to address, they're minor though.

VWoeltjen commented:
Addressed feedback from code review:

* Simplified logic when traversing objects to export to determine which columns to include; removed  ternaries and [defensive allocations](https://github.com/nasa/openmctweb/issues/732#issuecomment-195109712) to improve readability, more clearly separating examination of state (initializing variables) from logic about what to do with these states.
* Renamed TimelineCSVExporter to TimelineColumnizer to reflect its more modest role in the CSV export process

akhenry commented:
### Reviewer Checklist

1. Changes appear to address issue? __Y__
2. Appropriate unit tests included? __Y__
3. Code style and in-line documentation are appropriate? __Y__
4. Commit messages meet standards? __Y__
